Marksman/CodeActions.fs

```diff
@@ -0,0 +1,54 @@
+module Marksman.CodeActions
+
+open Ionide.LanguageServerProtocol.Types
+
+open FSharpPlus.GenericBuilders
+open Marksman.Toc
+open Marksman.Workspace
+open Marksman.Text
+open Marksman.Misc
+
+open type System.Environment
+
+type DocumentAction = { name: string; newText: string; edit: Range }
+
+let documentEdit range text documentUri : WorkspaceEdit =
+    let textEdit = { NewText = text; Range = range }
+
+    let mp = Map.ofList [ documentUri, [| textEdit |] ]
+
+    { Changes = Some mp; DocumentChanges = None }
+
+let tableOfContents (document: Doc) : DocumentAction option =
+    monad' {
+        let! toc = Toc.TableOfContents.mk document.index
```

Comment:
nit: since, this is the only binding let in this monad builder, perhaps a simple match would do? 

The problem with computation expression is that they are a bit harder to step through in the debugger compared to regular match - with.

Reply:
still relevant